java/maven/src/org/netbeans/modules/maven/options/NbArtifactVersionPreferences.java

```diff
@@ -0,0 +1,124 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.modules.maven.options;
+
+import java.util.Map;
+import java.util.prefs.Preferences;
+import org.apache.maven.artifact.versioning.ComparableVersion;
+import org.netbeans.modules.maven.api.Constants;
+import org.netbeans.modules.maven.indexer.api.NBVersionInfo;
+import org.netbeans.modules.maven.indexer.api.RepositoryQueries;
+import org.openide.util.NbPreferences;
+
+import static java.util.Map.entry;
+
+/**
+ * Preferences class for internal artifact version queries.
+ *
+ * Generally returns the latest known version unless overwritten by preference storage.
+ *
+ * @author mkleint
+ */
+public final class NbArtifactVersionPreferences {
+
+    private static final Map<String, String> fallback;
+
+    static {
+        // TODO update periodically - modifications might require test updates
+        String nb_version = "RELEASE200";
+        String nb_utilities_version = "14.0";
+        fallback = Map.ofEntries(
+            entry(key("org.netbeans.api", "org-netbeans-modules-editor"), nb_version), // represents all other nb artifacts
+            entry(key(Constants.GROUP_APACHE_PLUGINS, Constants.PLUGIN_COMPILER), "3.11.0"),
+            entry(key(Constants.GROUP_APACHE_PLUGINS, Constants.PLUGIN_RESOURCES), "3.3.1"),
+            entry(key(Constants.GROUP_APACHE_PLUGINS, Constants.PLUGIN_FAILSAFE), "3.2.2"),
+            entry(key(Constants.GROUP_APACHE_PLUGINS, Constants.PLUGIN_SUREFIRE), "3.2.2"),
+            entry(key("org.apache.netbeans.utilities", "utilities-parent"), nb_utilities_version),
+            entry(key("org.apache.netbeans.utilities", "nbm-maven-harness"), nb_utilities_version),
+            entry(key("org.apache.netbeans.utilities", "nbm-shared"), nb_utilities_version),
+            entry(key("org.apache.netbeans.utilities", "nbm-repository-plugin"), nb_utilities_version),
+            entry(key("org.apache.netbeans.utilities", "nbm-maven-plugin"), nb_utilities_version),
+            entry(key("org.apache.netbeans.archetypes", "nbm-archetype"), "1.18"),
+            entry(key("org.apache.netbeans.archetypes", "netbeans-platform-app-archetype"), "1.23")
+        );
+    }
+
+    private static final NbArtifactVersionPreferences INSTANCE = new NbArtifactVersionPreferences();
+
+    private NbArtifactVersionPreferences() {}
+
+    public static NbArtifactVersionPreferences getDefault() {
+        return INSTANCE;
+    }
+
+    protected final Preferences getPreferences() {
```

Comment:
i haven't changed the code, its from the first implementation which was not finished. I thought about removing it completely, but maybe having a way to override it easily might be useful for testing purposes.

what do you think, should I remove it instead of making it private? I presumed the fact that the class was final combined with the comments indicated that it was half finished and might be changed to non-final at some point. But maybe it was just a mistake.

Reply:
This comment belongs to all `protected` methods of this class - on a final class `protected final Preferences getPreferences(){}` and `final Preferences getPreferences(){}` have the effectively same visiblity.

Final classes are traditional in NetBeans for APIs, you don't risk anyone subclassing it and thus making additions of new methods dangerous.